### PR TITLE
OpenShift patches to Kustomize manifests

### PR DESCRIPTION
**Pull request by israel-hdez**
Related to opendatahub-io/odh-manifests#848

### CODE REVIEW

**Comment by ReToCode** [comment]
From the manifest-PR we also need:
- The env variable `ISTIO_WITH_CNI_AND_DNS_PROXY=true` in the `kserve-controller-manager`
- Add the [network-policy](https://github.com/opendatahub-io/odh-manifests/blob/e15bb8828d89c671cffaca8f70b03eb30b8e5833/kserve/odh-overlays/network-policies.yaml)

From Jooho's branch, seems like there is also some changes missing here:
- https://github.com/opendatahub-io/kserve/compare/master...Jooho:kserve:manifests#diff-a28989d682dc6ff9adc14f0c5da97da8dc74a0c6dfa1e0a58008a75cdd725aceR31
I think Jooho changes more stuff in his script. It think a good approach could be to compare the two folders `kserve` and `kserve_templates`. We should need all of those things in the diff. WDYT @jooho?

**Comment by ReToCode** [comment]
@Jooho you also have a [kfdef](https://github.com/opendatahub-io/kserve/compare/master...Jooho:kserve:manifests#diff-17f9efb2f583547c40ee50fa6a9118fb879c07b6b00f91810d911025f9e57dfaR1) in your `KServe` repo. Do you want to have it here or in the odh-manifests?

**Comment by skonto** [comment]
> The env variable ISTIO_WITH_CNI_AND_DNS_PROXY=true in the kserve-controller-manager

This has changed upstream we no longer use that check https://github.com/kserve/kserve/pull/2853.

The way to pass a uid is this:

```
kubectl apply -n kserve-demo -f - <<EOF
apiVersion: serving.kserve.io/v1beta1
kind: InferenceService
metadata:
  name: sklearn-from-uri
  annotations:
    sidecar.istio.io/inject: "true"
    sidecar.istio.io/rewriteAppHTTPProbers: "true"
    serving.knative.openshift.io/enablePassthrough: "true"
spec:
  predictor:
    annotations:
       serving.kserve.io/storage-initializer-uid: "1000700001"
    model:
      runtime: kserve-sklearnserver
      modelFormat:
        name: sklearn
      storageUri: https://github.com/skonto/serverless-guides/blob/main/kserve/kfserving-examples/models/sklearn/1.0/model/model.joblib?raw=true
EOF

```

**Comment by skonto** [comment]
so far LGTM

File: config/default/kustomization.yaml
Comment (Jooho) [inline config/default/kustomization.yaml:43]: do we need to delete this? we can add `service.beta.openshift.io/inject-cabundle: "true"` then the ca will be updated automatically. https://docs.openshift.com/container-platform/4.8/security/certificates/service-serving-certificate.html#add-service-certificate-mutating-webhook_service-serving-certificate

**Comment by Jooho** [comment]
@ReToCode We need the same manifest in odh-manifests this repo too for testing purposes. 
The process should be 
1. Sync upstream
2. Test a new manifests in the repo
3. Copy the final manifest into odh-manifests

Regarding kfdef, we don't need to put this into odh-manifests. This is also for testing purposes.

@israel-hdez I have a question. if we use the same branch strategy of modelserving, how we can have 2 different manifests for latest source and stable source?

**Comment by Jooho** [comment]
@skonto I discuss the 1337 userid issue with service mesh team and they recommend changing the way from init-container to sidecar way. I am not sure if kserve/kserve team accept this change or not but it does not seem that there are better options.

File: config/default/kustomization.yaml
Comment (israel-hdez) [reply config/default/kustomization.yaml:44]: These are here just because it replicates what we had in the `odh-manifests` PR.

That said, I'm happy to remove these patches if are unneeded.

**Comment by israel-hdez** [comment]
I added a commit rater than amending, for easier review). Once we are happy with this I can squash (or merge the individual commits, if you are OK).

> From the manifest-PR we also need:
> 
>     * The env variable `ISTIO_WITH_CNI_AND_DNS_PROXY=true` in the `kserve-controller-manager`
> 

From [Stavros comment](https://github.com/opendatahub-io/kserve/pull/14#issuecomment-1623094499) I understood this is not needed, so I omitted this. Please, let me know if I misunderstood.

>     * Add the [network-policy](https://github.com/opendatahub-io/odh-manifests/blob/e15bb8828d89c671cffaca8f70b03eb30b8e5833/kserve/odh-overlays/network-policies.yaml)
> 

I added the network policies in my last commit.

> 
> From Jooho's branch, seems like there is also some changes missing here:
> 
>     * [master...Jooho:kserve:manifests#diff-a28989d682dc6ff9adc14f0c5da97da8dc74a0c6dfa1e0a58008a75cdd725aceR31](https://github.com/opendatahub-io/kserve/compare/master...Jooho:kserve:manifests#diff-a28989d682dc6ff9adc14f0c5da97da8dc74a0c6dfa1e0a58008a75cdd725aceR31)
>       I think Jooho changes more stuff in his script. It think a good approach could be to compare the two folders `kserve` and `kserve_templates`. We should need all of those things in the diff. WDYT @Jooho?

I tried to inspect JooHo branch. Almost everything is already here, this is what looks missing:
* Removing the RBAC proxy in the manager. It is removed in my last commit.
* The ODH overlay includes kubeflow overlay here: https://github.com/Jooho/kserve/blob/manifests/opendatahub/odh-manifests/kserve_templates/kserve-controller/overlays/odh/kustomization.yaml#L5. Do we need this? I didn't add it, but if we do need the kubeflow overlay, I can do the change to this PR.
* I didn't include the scripts. If I understand correctly, that's mostly for manifests syncing between upstream and `odh-manifests`, and also for running tests. Just to keep work focused, I'm leaving that out from this PR. I'll include those in another PR, if needed.

**Comment by israel-hdez** [comment]
> @Jooho you also have a [kfdef](https://github.com/opendatahub-io/kserve/compare/master...Jooho:kserve:manifests#diff-17f9efb2f583547c40ee50fa6a9118fb879c07b6b00f91810d911025f9e57dfaR1) in your `KServe` repo. Do you want to have it here or in the odh-manifests?

@ReToCode I've seen that we have KfDefs here: https://github.com/opendatahub-io/odh-manifests/tree/master/kfdef. I use them just for reference, although I don't know if that's the intended goal. In any case, if we need the KfDef, my perception is that it would go in `odh-manifests` on that folder.

**Comment by israel-hdez** [comment]
> @israel-hdez I have a question. if we use the same branch strategy of modelserving, how we can have 2 different manifests for latest source and stable source?

@Jooho  From a developer point of view, my opinion is that the source code is self-contained in a commit/branch/tag.

If you need `latest`, you `git checkout master` and use the manifests you get under `/config`. If you need `stable`, you `git checkout v0.10.2` (or `release-0.10` or whatever version) and then, again, you use the manifests you get under `/config`.

Indeed, once this PR is merged, I plan to replicate what we have done here to `master` branch.

**Comment by ReToCode** [comment]
> Sync upstream
> Test a new manifests in the repo
> Copy the final manifest into odh-manifests

@Jooho yes, that should work with this PR and I updated mine in odh-manifests to just use the final files from this repo here.

> I discuss the 1337 userid issue with service mesh team(@bartoszmajsak @aslakknutsen ) and they recommend changing the way from init-container to sidecar way. I am not sure if kserve/kserve team accept this change or not but it does not seem that there are better options.

Hm sidecars are not run before the model server starts. It would need some kind of waiting logic in the server to let the sidecar finish. I doubt that KServe would want to change that. But feel free to start the discussion there.

> From https://github.com/opendatahub-io/kserve/pull/14#issuecomment-1623094499 I understood this is not needed, so I omitted this. Please, let me know if I misunderstood.

@israel-hdez It depends a bit on how we do the istio fix. It is not needed if the user provides the annotation himself on every InferenceService. I doubt that this is a feasible approach. I think we should patch the KServe fork to automatically set the runtime userId, then we would need a flag in ODH to enable this behaviour. So I would probably add this flag. WDYT @skonto?

>  I've seen that we have KfDefs here
and
> Regarding kfdef, we don't need to put this into odh-manifests. This is also for testing purposes.
@israel-hdez yeah, I was just wondering about the one Jooho added in his branch. Seems like it is there for testing and it would probably make sense to also have it here?

**Comment by skonto** [comment]
> It depends a bit on how we do the istio fix. It is not needed if the user provides the annotation himself on every InferenceService. I doubt that this is a feasible approach. I think we should patch the KServe fork to automatically set the runtime userId, then we would need a flag in ODH to enable this behaviour. So I would probably add this flag. WDYT @skonto?


Yes we need to have this automatically set to improve UX, already discussed [here](https://github.com/opendatahub-io/odh-manifests/pull/838#discussion_r1250813392).

**Comment by ReToCode** [comment]
> Yes we need to have this automatically set to improve UX, already discussed https://github.com/opendatahub-io/odh-manifests/pull/838#discussion_r1250813392.

Yep. The question now is just if we want to keep a flag to enable/disable this behaviour or not. If we always target OCP and SM we can also do it statically without a flag?

**Comment by skonto** [comment]
> If we always target OCP and SM we can also do it statically without a flag?
Yes that should be the case.
 
Note: Now we have to apply anyuid scc due to the python image uid issue. So even know it is not going to work out of the box for all models if the scc is not there.

> I am not sure if kserve/kserve team accept this change or not but it does not seem that there are better options.

Wrt to alternatives:

a)  the new sidecar feature coming as alpha (https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/753-sidecar-containers/README.md
https://github.com/knative/serving/issues/14023) in 1.28 might be an option as it introduces a new type of container that behaves like the init container but it does not block the pod progress.

Also I see a nice scenario:

```
Some pods use init containers to pull down configuration/secrets and update them before the main container gets it. Then use sidecars to continue to watch for changes and perform the updates and push to the main container. This requires two separate code paths today. Perhaps the same sidecar could be used for both cases.
```
Imagine the container downloading the model and keep watching its metadata for changes and probably download the new one and doing a swap in place.

b) The model server could download the model at its startup.

**Comment by ReToCode** [comment]
> Yes that should be the case.

Ok, then let's do that for now. @israel-hdez you can remove the env variable then. I'll add the code changes to KServe statically (for now).

> Note: Now we have to apply anyuid scc due to the python image uid issue. So even now it is not going to work out of the box for all models if the scc is not there.

That should be fixable by building those ourselves.

> Wrt to alternatives: 

Nice, I like your proposals. That way I can imagine bringing it also to Knative + KServe. But in the mean time we probably still need to go with the hack.

**Comment by israel-hdez** [comment]
> Ok, then let's do that for now. @israel-hdez you can remove the env variable then. I'll add the code changes to KServe statically (for now).

Ok! So, maybe this PR is ready?

Please, approve if so.

**Comment by israel-hdez** [comment]
> Ok! So, maybe this PR is ready?
> 
> Please, approve if so.

@ReToCode @skonto @vaibhavjainwiz  ^

**Comment by skonto** [comment]
Build is failing:

```
    1.6886675897979813e+09	INFO	V1beta1InferenceServiceController	Reconciling inference service	{"apiVersion": "", "isvc": "foo"}
    1.6886675898540587e+09	ERROR	V1beta1InferenceServiceController	Failed to reconcile	{"reconciler": "<*components.Predictor Value>", "Name": "foo", "error": "no matches for kind \"ClusterServingRuntime\" in version \"serving.kserve.io/v1alpha1\""}
```

**Review by vaibhavjainwiz — APPROVED** [review]

**Comment by israel-hdez** [comment]
> Build is failing:
> 
> ```
>    1.6886675897979813e+09	INFO	V1beta1InferenceServiceController	Reconciling inference service	{"apiVersion": "", "isvc": "foo"}
>     1.6886675898540587e+09	ERROR	V1beta1InferenceServiceController	Failed to reconcile	{"reconciler": "*components.Predictor Value>", "Name": "foo", "error": "no matches for kind \"ClusterServingRuntime\" in version \"serving.kserve.io/v1alpha1\""}
> ```
> 
> Is this expected?

@skonto I followed the guide at https://gist.github.com/skonto/5e3ac71cbb527eed629c8381edf1e425, and I didn't get the error in the first try.

I did a little additional testing. It turns out that when you create an `InferenceService`, if there is a suitable `ServingRuntime` in the namespace, all works OK. However, if there is no suitable `ServingRuntime`, then a lookup for a suitable `ClusterServingRuntime` is tried and, since the CRD is absent, this error is emitted. Relevant code block is [here](https://github.com/kserve/kserve/blob/release-0.10/pkg/controller/v1beta1/inferenceservice/utils/utils.go#L211-L225).

I did a quick scan of the code and I think this is the only error we will hit (at least, in the kserve-controller) if we omit the `ClusterServingRuntime` CRD. I think this error is not bad when compared with the unwanted implications of including this CRD. So, I'll leave things as is. I'll create a GitHub issue to patch our fork to minimally emit a better error message.

**Comment by ReToCode** [comment]
Couldn't we just have the CRD installed but without any instances of it? Only an admin could then add those. If we want to go further, we can prevent it with a Webhook. Would that help? 
I assume it is not ideal to have failing go builds on every PR here :)

**Comment by ReToCode** [comment]
> It can help. Although, if we anyway are going to change the code, isn't it better to fix the build? IDK.

I'd say it depends if you want to support ClusterRuntimes at some point or not. Changing KServe to support that, needs to be rebased/merged on every update. Adding a Webhook as an extension (does not have to be in KServe) could do that without changing KServe. We have to do the same thing for the explainers as well. /cc @danielezonca

**Comment by danielezonca** [comment]
> > It can help. Although, if we anyway are going to change the code, isn't it better to fix the build? IDK.
> 
> I'd say it depends if you want to support ClusterRuntimes at some point or not. Changing KServe to support that, needs to be rebased/merged on every update. Adding a Webhook as an extension (does not have to be in KServe) could do that without changing KServe. We have to do the same thing for the explainers as well. /cc @danielezonca

There is no plan at this stage to provide ClusterRuntimes support but I agree we should minimise the fork maintenance complexity: KServe upstream has this feature so even if we remove the code that uses it we will need to verify if everything it is still working for every release (it is possible that a new codepath will be added that assumes that CRD to exist and it will break ODH code).
I think we can raise this question to KServe community to see if they have any suggestion (aka we want to prevent the creation of ClusterRuntimes).
The Webhook extension (@ReToCode are you referring to [this k8s feature](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#validatingadmissionwebhook), right?) is definitely an option but it is not ideal that the CRD will be installed in the cluster so it might be misleading for the user.

@israel-hdez 
Wdyt? Is it fine for you to raise the question to KServe community? It should be ideal if a similar "flag" can be added to the project upstream so we can skip CRD installation and be sure it is not going to break at every sync

**Comment by israel-hdez** [comment]
> @israel-hdez
> Wdyt? Is it fine for you to raise the question to KServe community? It should be ideal if a similar "flag" can be added to the project upstream so we can skip CRD installation and be sure it is not going to break at every sync

It is a good way to pursue, IMO.
Although, I'm not sure if that should block this PR...

**Comment by danielezonca** [comment]
> > @israel-hdez
> > Wdyt? Is it fine for you to raise the question to KServe community? It should be ideal if a similar "flag" can be added to the project upstream so we can skip CRD installation and be sure it is not going to break at every sync
> 
> It is a good way to pursue, IMO. Although, I'm not sure if that should block this PR...
> 
> EDIT: Also, since this is `release-0.10`, any new flag may not be applied here. So, I think we need to fix the build in this branch and, in parallel, see if we can have discuss with upstream community the possibility to make `ClusterServingRuntime` optional in `master`.
> 
> But, as I said, I'm not sure if we should block this PR while we do all that.

I don't think we should block this PR, as intermediate fix I'm fine with any solution, changing that line of code or introduce the admission controllers, I think both are fine as fix until we agree the long term solution.

**Comment by israel-hdez** [comment]
> The E2Es aren't that happy. The error I'm seeing here is about some pod not starting. I'm not sure if it make sense to fix that in this PR. IMHO, that's work for another PR, given that we all are still getting familiar with this fork.

Indeed, updating E2Es are a different ticket in our backlog: https://github.com/opendatahub-io/kserve/issues/5

**Comment by israel-hdez** [comment]
@ReToCode @skonto I'm hoping this time I can have your approval here if all looks good for you, so that we can move forward.

**Review by ReToCode — APPROVED** [review]
I'd still vote not to change KServe to much for maintenance reasons. But lets try to find a solution for this upstream. So based on that and the discussion above, this is good with me:

/lgtm
/approve

**Pull request merged**
